File: go/leetcode/78.子集.go
/*
 * @lc app=leetcode.cn id=78 lang=golang
 *
 * [78] 子集
 */

// subsets 从空集开始，遍历时每次对结果集合内的每个结果继续叠加
// func subsets(nums []int) [][]int {
// 	ret := [][]int{}
// 	ret = append(ret, []int{})
// 	for i := 0; i < len(nums); i++ {
// 		for _, sub := range ret {
// 			sub = append(sub, nums[i])
// 			tmp := append([]int{}, sub...)
// 			ret = append(ret, tmp)
// 		}
// 	}
// 	return ret
// }

// 用回溯法，组合模式来做
func subsets(nums []int) [][]int {
	s := &solution{
		res:  [][]int{},
		path: []int{},
	}
	s.call(nums, 0)
	return s.res
}

type solution struct {
	path []int
	res  [][]int
}

func (s *solution) call(nums []int, start int) {
	if start > len(nums) {
		return
	}

	s.res = append(s.res, append([]int{}, s.path...))

	for i := start; i < len(nums); i++ {
		s.path = append(s.path, nums[i])
		s.call(nums, i+1)
		s.path = s.path[:len(s.path)-1]
	}
}

File: cpp/leetcode/78.子集.cpp
/*
 * @lc app=leetcode.cn id=78 lang=cpp
 *
 * [78] 子集
 *
 * https://leetcode-cn.com/problems/subsets/description/
 *
 * algorithms
 * Medium (79.88%)
 * Likes:    1214
 * Dislikes: 0
 * Total Accepted:    258.2K
 * Total Submissions: 323.2K
 * Testcase Example:  '[1,2,3]'
 *
 * 给你一个整数数组 nums ，数组中的元素 互不相同
 * 。返回该数组所有可能的子集（幂集）。
 *
 * 解集 不能 包含重复的子集。你可以按 任意顺序 返回解集。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：nums = [1,2,3]
 * 输出：[[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]
 *
 *
 * 示例 2：
 *
 *
 * 输入：nums = [0]
 * 输出：[[],[0]]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 1
 * -10
 * nums 中的所有元素 互不相同
 *
 *
 */

#include <vector>

// @lc code=start
class Solution {
private:
  std::vector<std::vector<int>> subsetsAppend(std::vector<int> &nums) {
    std::vector<std::vector<int>> ret;
    ret.push_back({});
    for (auto num : nums) {
      int size = ret.size();
      for (int i = 0; i < size; i++) {
        std::vector<int> arr = ret[i];
        arr.push_back(num);
        ret.push_back(arr);
      }
    }
    return ret;
  }

public:
  std::vector<std::vector<int>> subsets(std::vector<int> &nums) {
    return subsetsAppend(nums);
  }
};
// @lc code=end
